docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,551 @@
+	readOneOps := (*opMap)[OpTypeGet]
+	readManyOps := (*opMap)[OpTypeList]
+	updateOps := (*opMap)[OpTypeUpdate]
+	deleteOps := (*opMap)[OpTypeDelete]
+	getAttributesOps := (*opMap)[OpTypeGetAttributes]
+	setAttributesOps := (*opMap)[OpTypeSetAttributes]
+
+	for crdName, createOp := range createOps {
+		if m.cfg.IsIgnoredResource(crdName) {
+			continue
+		}
+		crdNames := names.New(crdName)
+...
+		// Now any additional Spec fields that are required from other API
+		// operations.
+		for targetFieldName, fieldConfig := range m.cfg.ResourceFields(crdName) {
+			if fieldConfig.IsReadOnly {
+				// It's a Status field...
+				continue
+			}
+
++			if fieldConfig.Reference != nil {
++				// Add the spec field here
++				referenceFieldNames := names.New(targetFieldName)
++				rf := NewReferenceField(crd, referenceFieldNames, fieldConfig)
++				crd.SpecFields[referenceFieldNames.Original] = rf
++				crd.Fields[referenceFieldNames.Camel] = rf
++			}
+
+...
+	m.crds = crds
+	return crds, nil
+}
+```
+
+```go
+//field.go
+func NewReferenceField(
+	crd *CRD,
+	fieldNames names.Names,
+	cfg *ackgenconfig.FieldConfig,
+) *Field {
+	gt := "*ackv1alpha1.AWSResourceReference"
+	gtp := "*ackv1alpha1.AWSResourceReference"
+	if cfg.Reference.IsList {
+		gt = "[]" + gt
+		gtp = "[]" + gtp
+	}
+	return &Field{
+		CRD:               crd,
+		Names:             fieldNames,
+		Path:              fieldNames.Original,
+		ShapeRef:          nil,
+		GoType:            gt,
+		GoTypeElem:        "AWSResourceReference",
+		GoTypeWithPkgName: gtp,
+		FieldConfig:       cfg,
+	}
+}
+```
+
+#### Runtime Changes
+
+```go
+// aws_resource_manager.go
+
+type AWSResourceManager interface {
+        ...
++       // DereferenceResource finds if there are any AWSResourceReference fields
++       // present in the AWSResource passed in the parameters and attempts to resolve
++       // those reference fields into target field.
++       // It returns an AWSResource with resolved references, boolean representing
++       // whether any reference fields were present and an error if the passed
++       // AWSResource's reference fields cannot be resolved
++       DereferenceResource(context.Context, client.Reader, AWSResource) (AWSResource, bool, error)
+}
+```
+
+```go
+// reconciler.go
+
+// reconcile either cleans up a deleted resource or ensures that the supplied
+// AWSResource's backing API resource matches the supplied desired state.
+//
+// It returns a copy of the resource that represents the latest observed state.
+func (r *resourceReconciler) reconcile(
+ctx context.Context,
+rm acktypes.AWSResourceManager,
+res acktypes.AWSResource,
+) (acktypes.AWSResource, error) {
+    if res.IsBeingDeleted() {
++       // Resolve references before deleting the resource.
++       // Ignore any errors while resolving the references
++    	  res, _, _ = rm.DereferenceResource(ctx, r.apiReader, res)
+        return r.deleteResource(ctx, rm, res)
+    }
+    return r.Sync(ctx, rm, res)
+}
```

Comment:
We can remove this from the proposal since it deals with something explicitly out-of-scope for the proposal (deletion propogation)

Reply:
This is needed to keep the existing deletion functionality. When reference is used, resolving the reference is required prior to deletion.
Ex: We need APIID to construct a DeleteIntegrationInput.